bootloader/lk.go

```diff
@@ -60,38 +81,189 @@ func (l *lk) Name() string {
+		//       on, since like the UC20 case, but that involves changing
+		//       many more tests, so let's do that in a followup PR
 		return filepath.Join(l.rootdir, "/dev/disk/by-partlabel/")
+	case RoleRecovery, RoleRunMode:
+		// TODO: maybe panic'ing here is a bit harsh...
+		panic("internal error: shouldn't be using lk.dir() for uc20 runtime modes!")
+	default:
+		panic("unexpected bootloader role for lk dir")
 	}
-	return filepath.Join(l.rootdir, "/boot/lk/")
 }
 
 func (l *lk) InstallBootConfig(gadgetDir string, opts *Options) error {
+	// make sure that the opts are put into the object
+	l.processOpts(opts)
 	gadgetFile := filepath.Join(gadgetDir, l.Name()+".conf")
-	systemFile := l.envFile()
+	systemFile, err := l.envFile()
+	if err != nil {
+		return err
+	}
 	return genericInstallBootConfig(gadgetFile, systemFile)
 }
 
 func (l *lk) Present() (bool, error) {
-	return osutil.FileExists(l.envFile()), nil
+	partitionLabelOrConfFile := l.partLabelForRoleAndTime()
+	// if we are not in runtime mode or in V1, just check the config file
+	if l.prepareImageTime || l.role == RoleSole {
+		return osutil.FileExists(filepath.Join(l.dir(), partitionLabelOrConfFile)), nil
+	}
+
+	// otherwise for V2, non-sole bootloader roles we need to check on the
+	// partition name existing, note that envFileForPartName will only return
+	// partiallyFound as true if it reasonably concludes that this is a lk
+	// device, so in that case forward err, otherwise return err as nil
+	_, partiallyFound, err := l.envFileForPartName(partitionLabelOrConfFile)
+	if partiallyFound {
+		return partiallyFound, err
+	}
+	return false, nil
 }
 
-func (l *lk) envFile() string {
+func (l *lk) partLabelForRoleAndTime() string {
+	// TODO: should the partition labels be fetched from gadget.yaml instead? we
+	//       have roles that we could use in the gadget.yaml structures to find
+	//       them
+	label := ""
+	switch l.role {
+	case RoleSole, RoleRunMode:
+		label = "snapbootsel"
+	case RoleRecovery:
+		label = "snaprecoverysel"
+	default:
+		panic(fmt.Sprintf("unknown bootloader role for littlekernel: %s", l.role))
+	}
+	if l.prepareImageTime {
+		// conf files at build time have .bin suffix, so technically this now
+		// becomes not a partition label but a filename, but meh names are hard
+		label += ".bin"
+	}
+	return label
+}
+
+func (l *lk) envFile() (string, error) {
 	// as for dir, we have two scenarios, image building and runtime
-	if l.inRuntimeMode {
-		// TO-DO: this should be eventually fetched from gadget.yaml
-		return filepath.Join(l.dir(), "snapbootsel")
+	partitionLabelOrConfFile := l.partLabelForRoleAndTime()
+	if l.prepareImageTime {
+		return filepath.Join(l.dir(), partitionLabelOrConfFile), nil
+	}
+
+	if l.role == RoleSole {
+		// see TODO: in l.dir(), this should eventually also be using
+		// envFileForPartName() too
+		return filepath.Join(l.dir(), partitionLabelOrConfFile), nil
+	}
```

Comment:
Can we do `if l.prepareImageTime || l.role == RoleSole` like we do in Present()?

Reply:
maybe not, if we consider my issues with if prepareImageTime in partLabelForRoleAndTime

Reply:
I think this is better now, let me know what you think